### PR TITLE
feat(hooks): add a new hook named useExternal

### PR DESCRIPTION
**Pull request by turkyden**
哈喽, 之前对于新增 `hooks/useExternal` 的 [RFC] 提议 #758 已通过评审，现在提个 PR，望采纳，感谢！ 

@brickspert

### CODE REVIEW

File: packages/hooks/src/useExternal/index.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  load: (newPath?: string) => void;
+  unload: () => void;
+};
+
+export type ExternalElement = HTMLScriptElement | HTMLLinkElement | HTMLImageElement;
+
+export default function useScript(path: string, options?: Options): [Status, Action] {
+  const [status, setStatus] = useState<Status>(path ? 'loading' : 'unset');
+
+  const [src, setSrc] = useState(path);
```
Comment (brickspert) [inline packages/hooks/src/useExternal/index.ts]: path 和 src 两个属性有点冗余了。

现在提供了两种方案来修改资源路径，一个是 path 变化，一个是 load/toggle 函数。
建议只保留 path，不要 src 了。也就是 load 和 toggle 的时候不再接收参数。

Comment (turkyden) [reply packages/hooks/src/useExternal/index.ts]: 这一点我也纠结了很久，OK, 那我改一改再提 PR 哈，感谢反馈！

**Comment by turkyden** [comment]
哈喽，已经修改 08ce045，再次 PR ！@brickspert

**Review by brickspert — APPROVED** [review]
LGTM

File: packages/hooks/src/useExternal/index.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+export type Action = {
+  toggle: () => void;
+  load: () => void;
+  unload: () => void;
+};
+
+export type ExternalElement = HTMLScriptElement | HTMLLinkElement | HTMLImageElement;
+
+export default function useScript(path: string, options?: Options): [Status, Action] {
```
Comment (awmleer) [inline packages/hooks/src/useExternal/index.ts]: 这里的函数名为啥是 `useScript`

File: packages/hooks/src/useExternal/index.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      }
+    };
+  }, [path, active]);
+
+  const action = useMemo(() => {
+    const unload = () => setActive(false);
+    const load = () => setActive(true);
+    const toggle = () => setActive(!active);
+    return { toggle, load, unload };
+  }, [active]);
```
Comment (awmleer) [inline packages/hooks/src/useExternal/index.ts]: 这里是不是可以这样优化一下

```suggestion
  const action = useMemo(() => {
    const unload = () => setActive(false);
    const load = () => setActive(true);
    const toggle = () => setActive(val => !val);
    return { toggle, load, unload };
  }, [setActive]);
```

File: packages/hooks/src/useExternal/index.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+export type ExternalElement = HTMLScriptElement | HTMLLinkElement | HTMLImageElement;
+
+export default function useScript(path: string, options?: Options): [Status, Action] {
+  const isPath = typeof path === 'string' && path !== '';
+
+  const [status, setStatus] = useState<Status>(isPath ? 'loading' : 'unset');
+
+  const [active, setActive] = useState(isPath);
+
+  const ref = useRef<ExternalElement>();
```
Comment (awmleer) [inline packages/hooks/src/useExternal/index.ts:27]: 这里为啥要用一个 ref，可以直接在下面的 `useEffect` 中用一个局部变量实现么

**Review by awmleer — CHANGES_REQUESTED** [review]

File: packages/hooks/src/useExternal/index.ts (hunk 2 above)
Comment (turkyden) [reply packages/hooks/src/useExternal/index.ts]: 噢，我统一成 `useExternal` 吧。

File: packages/hooks/src/useExternal/index.ts (hunk 3 above)
Comment (turkyden) [reply packages/hooks/src/useExternal/index.ts]: 确实这样更优雅！我改一下哈。

File: packages/hooks/src/useExternal/index.ts (hunk 4 above)
Comment (Turkyden) [reply packages/hooks/src/useExternal/index.ts:27]: 这里的 ref 我考虑是用于存储当前外链资源的 dom 实例，如果不使用 ref，当 path 是动态改变的时候，useExternal 将无法移除之前旧的 dom. @awmleer

Comment (Turkyden) [reply packages/hooks/src/useExternal/index.ts:27]:
```js
const [ status, toggle ] = useExternal(path, {})
```
path 如果动态传值的话，如果不用 ref 存储住外链 dom，貌似无法移除旧的外链 dom。

Comment (turkyden) [reply packages/hooks/src/useExternal/index.ts:27]: > 这里为啥要用一个 ref，可以直接在下面的 `useEffect` 中用一个局部变量实现么

这里的 ref 我是这样考虑的，它的作用是存储当前外链资源的 DOM 引用。

```js
const [ status, { load, unload, toggle } ] = useExternal(path, options);
```

如果不用 ref 存储住外链资源的 DOM 引用，当参数 `path`  是动态传入时，`useExternal` 貌似无法移除之前加载的外链 DOM，然后替换新的外链资源。

```diff
-<script src="foo.js" async data-status="ready"></script>
+<script src="baz.js" async data-status="ready"></script>
```

我暂时还想不到更好的解决办法，因此，`useEffect` 中使用局部变量貌似无法实现这个需求。 

@awmleer

**Comment by ouzhou** [comment]
为啥不叫use script

**Comment by turkyden** [comment]
> 为啥不叫use script

因为不仅仅用于加载 js 哦，还可以加载 css, img 这些外链资源。

受启发于 [codesandbox](https://codesandbox.io/) 的 External Resources 功能。

**Comment by ouzhou** [comment]
看起来似乎是CI在拉取git的时候报错了 @Turkyden 
很期待这个hook能使用

**Comment by turkyden** [comment]
> 看起来似乎是CI在拉取git的时候报错了 @Turkyden
> 很期待这个hook能使用

是的，这个新 hook 在等待 @awmleer 的 review 哦。

**Review by awmleer — APPROVED** [review]
LGTM

**Comment by KrishnaPG** [comment]
This is a good starting point, and very useful to inject dynamic dependencies from CDN. However, few important pieces are missing in the functionality:

1. Duplicates should be avoided. For example, multiple instances of a component injecting the same CSS or Script should be handled
2. `useExternal` should be able to add more than one script/css at a time. That is, should support array of URLs in case more than one is to be injected
3. Unloading the script/css should be taken care of.

It would be a great value if these above also could be addressed.

**Comment by brickspert** [comment]
> This is a good starting point, and very useful to inject dynamic dependencies from CDN. However, few important pieces are missing in the functionality:
> 
> 1. Duplicates should be avoided. For example, multiple instances of a component injecting the same CSS or Script should be handled
> 2. `useExternal` should be able to add more than one script/css at a time. That is, should support array of URLs in case more than one is to be injected
> 3. Unloading the script/css should be taken care of.
> 
> It would be a great value if these above also could be addressed.

Yeah, your suggestion is very good, thank you.

@TurkydenCan you support these features？

File: packages/hooks/src/useExternal/index.ts (hunk 3 above)
Comment (wangcongyi) [reply packages/hooks/src/useExternal/index.ts]: 这里为啥使用 `setActive` 作为依赖项?

**Comment by iShawnWang** [comment]
> This is a good starting point, and very useful to inject dynamic dependencies from CDN. However, few important pieces are missing in the functionality:
> 
> 1. Duplicates should be avoided. For example, multiple instances of a component injecting the same CSS or Script should be handled
> 2. `useExternal` should be able to add more than one script/css at a time. That is, should support array of URLs in case more than one is to be injected
> 3. Unloading the script/css should be taken care of.
> 
> It would be a great value if these above also could be addressed.

Awesome suggestions !!! 👍

**Pull request merged**